### ISSUE TITLE
Color Hex transformations to top surfaces are illogical

### ISSUE THREAD

**Issue by nopunin10did** [ijprest/keyboard-layout-editor#283]
### Brief Summary
When KLE currently simulates a keycap in a given color, it takes the user-provided color hex and uses that value for the base of the keycap.  However, for the top surface of the cap, it uses a transformed color.  This in itself isn't a bad thing, but the transformation formula currently in use appears to have several inconsistencies.
To be clear, this is not a question of whether KLE looks "realistic". While I have no expectation that the KLE simulations would look like the actual colors of a keyset (given that RGB is different that physical colors, yada yada), it would make sense for the top-surface transformations to be consistent at least from a digital / mathematical standpoint.

### Expected Behavior
One of these two formulas (or something similar) should be applied to every cap:
1) Top surfaces should be the original color hex provided by the user, with base/sides of the caps being a shade darker. This would approximate neutral white light on a key's top surface with shadows on the sides.
2) Base/sides of the caps should be the original color hex provided by the user, with the top of the cap being a shade lighter. This would approximate brighter white light that reduces shadows on the sides but lightens (and slightly washes out) the tops.

In either case, the only transformations should be: 
(A) Change the lightness of the base color (lower for formula 1, higher for formula 2) 
(B) Change the saturation of the base color (always lower, regardless of formula 1 or 2). 

**Saturation should never increase** in a transformation of this sort, and hue should never change (unless the color is a shade of pure white/grey/black in which case hue is irrelevant).

### Actual Behavior
It appears that KLE is attempting to take approach (2) from above.  However, for several colors, the transformation is actually **increasing** the saturation of the color or (in some cases) making significant changes to the hue of the color.

### Examples
Attached is a spreadsheet that converts the two hexes (original and top) provided by KLE for certain colors into an HSL basis.

[KLE color bug notes.xlsx](https://github.com/ijprest/keyboard-layout-editor/files/4880074/KLE.color.bug.notes.xlsx)

In this spreadsheet, H is represented by angle in degrees (0-360 scale), while all other values (R, G, B, S, & L) are represented on a 0-1 scale.

_Example of Expected Behavior_
A user-supplied hex of **e49e00** yields a top hex of **ffc53d**.  This is fine, as the saturation is at 1.0 in both cases and the hue is not altered (within acceptable rounding error).  The top surface is about 0.2 lighter, which is also fine.

_Example of Inappropriately Increased Saturation_
A user-supplied hex of **d2aa6d** yields a top hex of **fed394**.  This is just bizarre, as it increases the saturation by over 0.4.  Lightness is increased as expected, and hue is only changed within an acceptable rounding error, but simulated lighting should not increase color saturation in that fashion.

_Example of Inappropriately Altered Hue_
A user-supplied hex of **faca30** yields a top hex of **fff862**.  Here, the saturation is increased slightly, which ideally shouldn't occur, but the more glaring error is that the hue of the original color is changed by nearly 12 degrees.

**Comment by vomindoraan** [comment]
Where in the code exactly is this color calculation performed?

**Comment by nopunin10did** [comment]
Thanks for finding that @vomindoraan ; it appears the problem here is that a trivial application of a percentage modifier to the L component of an LAB color doesn't reliably preserve the hue of that color.

I think part of the issue may be that LAB was designed to model differences in color _perception_, and while the L component is analogous to a lightness component, increasing L while keeping A and B constant doesn't actually preserve hue the way one would expect.

I'm not sure what a trivial replacement would be, though.  For one thing, a 20% multiplicative increase in L across the board doesn't seem like a great approximation of a constant neutral light source over similar materials of different colors.  There's got to be a better way to handle that.

**Comment by nopunin10did** [comment]
Maybe the better approach here would be to leave the hex of the key tops intact (i.e. use the user-provided hex, as it currently does with text) and just darken the key sides instead.  I'll play around with some math on this.

**Comment by vomindoraan** [comment]
The sRGB ↔ L*a*b* transformations aren't perfect either, there are losses which result in slightly different colors coming out on the other end. It would be better to convert to HSL instead, or scale the RGB values directly.

**Comment by vomindoraan** [comment]
Although, to be fair, human vision is not linear, and perhaps we are wrong in assuming that this transformation is wrong/illogical. It may look more accurate to the human eye than a straight up lightness increase would.

**Comment by vomindoraan** [comment]
Either way, if someone comes across this issue looking for a way to calculate which color they need to type into the color boxes in order to get the desired color (approximately) on the key tops, this function will do that:

```js
function darkenColor(hex) {
  let c = $color.hex(hex).Lab();
  c.l /= 1.2;
  return c.sRGB8().hex();
}
```

You can run it in your browser's console (F12) on any KLE page. It takes as an argument a hex color string (e.g. `"#818283"`) and returns the result in the same format. Example usage:

```js
→ darkenColor("#818283");
← "#6b6c6d"
```

**Comment by nopunin10did** [comment]
> Although, to be fair, human vision is not linear
This is true, but what I'm assuming is intended here is a change to the surface that simulates a difference in lighting using a neutral, distant light source.  If we were looking at this like a trivial 3D render, then the key tops would reflect both their diffuse and ambient material colors at full intensity while the key bottoms/sides would reflect at full ambient intensity but only a portion of diffuse.

Such lighting models generally don't light according to LAB but rather just keep everything in sRGB or whatnot.  Therefore, a constant light across multiple colors should mathematically alter all of them in a similar fashion, assuming the underlying materials are also similar. And the _perceived_ difference in lightening could vary from key to key.

I think what might be in order here as a trivialized version of this notion would be to
1) Use the base hex as the color of the key top.
2) Convert that hex to HSL rather than LAB.
3) Multiple the L component by (1/1.2).
4) Convert the new HSL to a new hex (hex2).
5) Apply that hex to the key bottoms/sides.

**Comment by mjsadaway** [comment]
As a workaround, https://kle-render.herokuapp.com/ gives a more realistic approximation of the true colors.

**Comment by nopunin10did** [comment]
Yes, it is better at displaying colors because it uses a standard 3D lighting calculation. That's not really an option here, but it would be good to approximate a similar result.

One thing to note is that 3D approximations of lighting generally don't change according to an LAB model but rather tend to just stick with RGB the whole time.  The actual rendered RGB / HSL values of various colored cap surfaces under a simplistic lighting model should be modified either similarly or proportionally assuming that their underlying material color values (ambient v. diffuse v. specular) are similar in proportion if not in actual value.  However, if you were to use LAB/XYZ conversions to compare lit-rendered-cap-versus-base for those different colors, the differences from color-to-color would be all over the map.  

There's not going to be consistency in the perceived difference in color when lit, though there should be some consistency in the underlying color when lit.

I think I may have a conversion in mind that might get us closer to something render-looking.
1. Convert the base color RGB(0) to HSL(0) (H expressed as 0-360, S & L expressed as 0-100).
2. Let the L component of HSL(0) be L(0).
3. Let L(1) = L(0) / mod.  (With mod being 1.2 as before, or possibly something else).
4. Let H(1) = H(0) and S(1) = S(0).
5. Create a new HSL color HSL(1) from H(1), S(1), and L(1).
6. Convert HSL(1) to RGB(1).
7. Use RGB(0) for the keycap tops. Use RGB(1) for the keycap sides.

**Comment by nopunin10did** [comment]
Ach, I just realized I'm repeating myself.  Whoops!

**Comment by mjsadaway** [comment]
No worries. Redundancy can be good, and, based on my limited knowledge of such things, your method sounds good. It would be better if we could test it. The real problem is that ijprest doesn't seem to be willing and able to address the problem.

As it is, KLE by itself is inadequate for designing color schemes. A shame, since it's otherwise pretty darn nice and could be taken to the next level if only.
